### PR TITLE
Use unreachable exception in favor of type safety

### PR DESCRIPTION
**Pull request by fa93hws**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 👻 What's the background?
`Switch` is not type safe.
e.g.
```
type Pet = 'dog' | 'cat';

function getPet(pet: Pet) {
  switch(pet) {
    case 'dog':
        return new Dot();
    case 'cat':
        return new Cat();
    default:
        throw new Error('unknown pet');
  }
}
```
Which looks good now.

However, if another value is added to the enum, e.g. `type Pet = 'dog' | 'cat' | 'hamster'`.
The `getPet` function stay silent in the compile time.
It may (or may not) throw an error in the run time which is error prone.

### 💡 Solution
It is possible to utilize the `never` type in typescript to force the switch statement to behave like an exhaustive pattern match to enhance the type safety.

B6A8D1EA-D2E6-4901-BE05-FC9A526B9C57

All occurrence will be changed if the gist of this PR is preferred.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by fa93hws** [comment]
> Could you please add tests to make sure this change works as expected?

I don't think we need test for this class...
Basically it's a functionality provided by the Typescript.
And I believe there has already been massive and reliable test on it.

There isn't much value to test the text thrown is correct or not....

File: components/button/button-group.tsx [hunk 1]
```diff
@@ -25,8 +26,11 @@ const ButtonGroup: React.SFC<ButtonGroupProps> = props => (
           break;
         case 'small':
           sizeCls = 'sm';
-        default:
+        case 'default':
+        case undefined:
           break;
+        default:
+          throw new UnreachableException(size);
```
Comment (fa93hws) [inline components/button/button-group.tsx]: `throw` could too aggressive for a library, maybe `console.error`?

Comment (afc163) [reply components/button/button-group.tsx]: use `warning`

**Comment by afc163** [comment]
Could you add a test case?

File: components/button/button-group.tsx (hunk 1 above)
Comment (afc163) [reply components/button/button-group.tsx]: ping @fa93hws

**Comment by fa93hws** [comment]
Sorry, looks like I've done something wrong so that the PR is closed automatically...
I create another one in https://github.com/ant-design/ant-design/pull/22933

**Pull request closed without merging**